src/test/java/com/emc/object/s3/S3JerseyClientTest.java

```diff
@@ -2971,4 +2973,182 @@ protected void assertAclEquals(AccessControlList acl1, AccessControlList acl2) {
+        client.deleteObjectTagging(new DeleteObjectTaggingRequest(bucketName, key).withVersionId(versionId));
+        t = client.getObjectTagging(new GetObjectTaggingRequest(bucketName, key).withVersionId(versionId));
+        Assert.assertNull(t.getTagSet());
+
+    }
+
+    @Test
+    public void testGetPutDeleteObjectWithTagging() {
+
+        // set up env
+        String bucketName = getTestBucket(), key = "test-object-tagging";
+        client.setBucketVersioning(bucketName, new VersioningConfiguration().withStatus(VersioningConfiguration.Status.Enabled));
+        client.putObject(new PutObjectRequest(bucketName, key, "Hello Version 1 !")
+                .withObjectTagging(new ObjectTagging().withTagSet(Arrays.asList(new ObjectTag("k11", "v11"), new ObjectTag("k12", "v12")))));
+        String versionId1 = client.listVersions(bucketName, null).getVersions().get(0).getVersionId();
+        client.putObject(new PutObjectRequest(bucketName, key, "Hello Version 2 !"));
+        String versionId2 = client.listVersions(bucketName, null).getVersions().get(0).getVersionId();
+
+        // Only the particular version of the object should get deleted and no other versions of object should be affected
+        client.deleteObject(new DeleteObjectRequest(bucketName, key).withVersionId(versionId2));
+        Assert.assertEquals(2, client.getObject(new GetObjectRequest(bucketName, key).withVersionId(versionId1), String.class).getObjectMetadata().getTaggingCount());
+
+        // Object and associated multiple tags should get deleted
+        Assert.assertEquals(2, client.getObject(new GetObjectRequest(bucketName, key).withVersionId(versionId1), String.class).getObjectMetadata().getTaggingCount());
+        client.deleteObject(new DeleteObjectRequest(bucketName, key).withVersionId(versionId1));
+        try {
+            client.getObjectTagging(new GetObjectTaggingRequest(bucketName, key).withVersionId(versionId1));
+            Assert.fail("Fail was expected. Can NOT get tags from a deleted object");
+        } catch (S3Exception e) {
+            Assert.assertEquals(404, e.getHttpCode());
+            Assert.assertEquals("NoSuchKey", e.getErrorCode());
+        }
+
+    }
+
+    @Test
+    public void testCopyObjectWithTagging() {
+
+        // set up env
+        String bucketName = getTestBucket(), key1 = "test-object-tagging-src", key2 = "test-object-tagging-dest1", key3 = "test-object-tagging-dest2", content = "Hello Object Tagging!";
+
+        // should be able to copy the object and copied object should have the tags also
+        client.putObject(new PutObjectRequest(bucketName, key1, content)
+                .withObjectTagging(new ObjectTagging().withTagSet(Arrays.asList(new ObjectTag("k11", "v11"), new ObjectTag("k12", "v12")))));
+        client.copyObject(new CopyObjectRequest(bucketName, key1, bucketName, key2));
+        client.copyObject(new CopyObjectRequest(bucketName, key1, bucketName, key2));
```

Comment:
should the target here be `key3`?  otherwise, where is key3 used?

Reply:
changed. key2 is for un-versioned object copying, key3 is for another versioned one.